overlays/bootstrap.nix

```diff
@@ -497,7 +523,16 @@ in {
 
     # the bootstrap infra structure (pre-compiled ghc; bootstrapped cabal-install, ...)
     bootstrap = with final.haskell-nix;
-      let ghc = final.buildPackages.haskell-nix.bootstrap.compiler."${buildBootstrapper.compilerNixName}";
+      let
+        # This compiler-nix-name will only be used to build nix-tools and cabal-install
+        # when checking materialization of alex, happy and hscolour.
+        compiler-nix-name = "ghc865";
```

Comment:
Do we *need* to build `nix-tools` and `cabal-install` with the user's GHC? Both of them should be fairly compiler-agnostic, the most relevant thing there will probably be what version of cabal they use. 

I think we could plausibly have an `internalCompilerNixName` for building such things. Then we wouldn't have to have a whole set of nix-tools expressions, wouldn't have to pass the compiler version into a bunch of the call-cabal-to-nix stuff.

Reply:
> Do we need to build nix-tools and cabal-install with the user's GHC?

I don't think we need to, but it has at least two nice advantages:

* Less overhead on macOS where dynamic linking pulls in the compiler used for nix-tools and cabal-install.
* Tests that haskell.nix can making it easy to compile stuff with different ghcs.

Reply:
The main thing is that we're spending quite a lot of complexity on threading compiler versions through when we don't actually need them.

And I believe we can compile things with different GHCs - we have lots of tests for that!

Reply:
Did we resolve this? I think this was my main gripe. If you want to do it this way it's fine, I just think it's causing a lot of suffering for a comparatively minor gain.

(Not to mention, if we used a fixed GHC for nix-tools we'd get better caching)

Reply:
Its only really a problem for stack projects (for cabal projects we always have a compiler-nix-name ahead of time).  For stack projects (and cabal projects) we use `config.compiler.nix-name` once it is available (for `shellFor { tools = ... }` and `p.tool`.  That only leaves `nix-tools` for use in the `stackProject` stuff.  For that I have added `internalDefaultCompilerNixName`.  In theory someone could add an overlay to set it, but it is only used to selecting the `nix-tools` build to use for stack projects (so hopefully no one will need to).  It is currently set to `"ghc883"`.

Reply:
I'm don't think it's a problem for users, I think it's a problem for *us*! At the moment AFAICT a significant fraction of the compiler name threading in this PR is *just* to get nix-tools... which is a set of binaries that don't have to be built with the user's compiler! So why are we doing that?

To put it another way: if we think of the nix-tools *binaries* as part of the haskell.nix interface, it's perfectly fine for them to be built with an `internalCompilerNixName`, so long as that doesn't leak out.

Reply:
> AFAICT a significant fraction of the compiler name threading in this PR is just to get nix-tools

We have to have `compiler-nix-name` already in `cabalProject` (to know what GHC to compile the project with).  The threading in the PR is mostly for that.

Reply:
Okay, I'm not sure how much of it we'd get rid of (at least various things like the updater scripts now need compiler names where they didn't before), but I do just think it's simpler. There is certainly a lot of indexing into this set of `nix-tools` derivations that would be more straightforward if there were only one!

I guess I also still don't see what the advantage is. Just the dynamic linking thing?

Reply:
> There is certainly a lot of indexing into this set of nix-tools derivations that would be more straightforward if there were only one!

I'll replace `internalDefaultCompilerNixName` with `internal-nix-tools` and also use it in the scripts.

> I guess I also still don't see what the advantage is. Just the dynamic linking thing?

Yes that is my main concern.  On macOS it is likely build or download a whole extra GHC.  If the user tries to solve that by passing in your own `nix-tools` to `cabalProject` but then everything in the cache that depends on `nix-tools` does not match.

Reply:
Okay: I think this discussion deserves a note in the code explaining the reasoning, since it's a bit non-obvious in the end.